### PR TITLE
Add tox.ini

### LINKED ISSUE #2010: Add tox.ini instead of checks.sh
## Motivation

I think tox is a better solution of [checks.sh](https://github.com/optuna/optuna/blob/master/checks.sh).
tox makes it easier to run tests and lint checks for newcomers because it is a widely accepted as a standard Python tool.

## Description

The setting file (tox.ini) of [optuna-dashboard](https://github.com/c-bata/optuna-dashboard) project is here:

```
[tox]
envlist =
    py36
    py37
    py38
    py39
    flake8
    black
    mypy

[testenv]
basepython = python3.9
commands = python -m unittest

[testenv:py36]
basepython = python3.6

[testenv:py37]
basepython = python3.7

[testenv:py38]
basepython = python3.8

[testenv:flake8]
deps = flake8
commands = flake8 .

[testenv:black]
deps = black
commands = black --check .

[testenv:mypy]
deps = mypy
commands = mypy .
```

Usage is like this:

```
$ pip install tox
$ tox -e ALL  # run all tests and lints
...
  py36: commands succeeded
  py37: commands succeeded
  py38: commands succeeded
  py39: commands succeeded
  flake8: commands succeeded
  black: commands succeeded
  mypy: commands succeeded
  congratulations :)
```

```
$ pip install tox
$ tox -e flake8 -e black -e mypy  # run lints
...
  flake8: commands succeeded
  black: commands succeeded
  mypy: commands succeeded
  congratulations :)
```

See the tox documentation for more details.

https://tox.readthedocs.io/en/latest/

### PR DESCRIPTION
**Pull request by c-bata**
## Motivation

To close #2010.

tox makes it easier to run tests and lint checks for newcomers because it is a widely accepted as a standard Python tool.

## Description of the changes

Add tox.ini, the setting file of tox.

```
$ tox -e flake8 -e black -e blackdoc -e mypy
...
  flake8: commands succeeded
  black: commands succeeded
  blackdoc: commands succeeded
  mypy: commands succeeded
  congratulations :)
```

```
$ tox -e test-py38
...
  test-py38: commands succeeded
  congratulations :)
```

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    blackdoc
+    mypy
+
+[testenv:test-py36]
+basepython = python3.6
+extras =
+    tests
+    optional
+deps =
+    scikit-optimize
```
Comment (HideakiImamura) [inline tox.ini:17]: We don't need `scikit-optimize` except for integration tests.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    integration
+commands = pytest -s tests/integration_tests
+
+[testenv:flake8]
+deps = flake8
+commands = flake8 .
+
+[testenv:black]
+deps = black
+commands = black --check .
```
Comment (HideakiImamura) [inline tox.ini:85]: How about adding the `--diff` option as [here](https://github.com/optuna/optuna/blob/master/.github/workflows/checks.yml#L26)?

File: tox.ini [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+basepython = python3.9
+extras =
+    tests
+    integration
+commands = pytest -s tests/integration_tests
+
+[testenv:flake8]
+deps = flake8
+commands = flake8 .
+
```
Comment (HideakiImamura) [inline tox.ini:82]: The `isort` block is missing.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! I think it's a good idea to maintain `tox` to make it easier for developers who are familiar with Python to participate in Optuna development.

I tried to use `tox` in my local environment, but I have the following errors. I don't have python 3.6, 3.7, or 3.9 in my environment, so I seem to get errors. Is this inevitable? (We don't normally have multiple versions of Python in the environment, right?)
スクリーンショット 2020-11-16 14 12 41

**Comment by c-bata** [comment]
> Thanks for the PR! I think it's a good idea to maintain `tox` to make it easier for developers who are familiar with Python to participate in Optuna development.
> 
> I tried to use `tox` in my local environment, but I have the following errors. I don't have python 3.6, 3.7, or 3.9 in my environment, so I seem to get errors. Is this inevitable? (We don't normally have multiple versions of Python in the environment, right?)
> スクリーンショット 2020-11-16 14 12 41

Yes. You need to install Python 3.6, 3.7 and 3.9 and add to $PATH.

I installed all python versions like below:

```
(venv) $ which python2.7
/Users/c-bata/bin/python2.7
(venv) $ which python3.6
/Users/c-bata/bin/python3.6
$ which python3.7
/Users/c-bata/bin/python3.7
$ which python3.8
/Users/c-bata/bin/python3.8
$ which python3.9
/Users/c-bata/src/github.com/optuna/optuna/venv/bin/python3.9
```

File: tox.ini (hunk 1 above)
Comment (c-bata) [reply tox.ini:17]: I add scikit-optimize due to the following comment:
https://github.com/optuna/optuna/blob/fbe4e6160fddf811659ad6d34ad7658db1f3ff6c/.github/workflows/tests.yml#L66-L69

Is it no longer needed?

Comment (c-bata) [reply tox.ini:17]: I tried to remove scikit-optimize, then unit tests are failed. It seems required for now.

**Comment by c-bata** [comment]
Thank you for your review! I applied your suggestions. PTAL.

File: tox.ini (hunk 1 above)
Comment (HideakiImamura) [reply tox.ini:17]: Sorry for the confusion. I missed the test case. You are right. Let me resolve this conversation.

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,93 @@
+[tox]
+envlist =
+    test-{py36,py37,py38,py39}
+    integration-test-{py36,py37,py38,py39}
```
[the remark this answers is not in the document]
Comment (HideakiImamura) [reply tox.ini]: `test-py39` is still existing. Could you also remove it?

[the remark this answers is not in the document]
Comment (c-bata) [reply tox.ini]: Oh, sorry. I push https://github.com/optuna/optuna/pull/2024/commits/2a7886a33b92c1726126a03bd7c72488f5bb855a to remove test-py39.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the swift actions. LGTM!

**Comment by HideakiImamura** [comment]
@c-bata I think the configuration of `tox` can be also used for CI (GitHub Actions). What do you think?

@crcrpar Do you have any comments on this PR?

**Comment by c-bata** [comment]
> @c-bata I think the configuration of tox can be also used for CI (GitHub Actions). What do you think?

tox conflicts with the matrix functionality of GitHub Actions and there are no benefits to create virtualenv. So I think it's better to use tox only for local machines.

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+deps = isort
+commands = isort . --check --diff
+
+[testenv:black]
+deps = black
+commands = black . --check --diff
+
+[testenv:blackdoc]
+deps = blackdoc
+commands = blackdoc . --check
```
Comment (c-bata) [inline tox.ini:76]:
```suggestion
commands = blackdoc . --check --diff
```

See https://github.com/optuna/optuna/pull/2031/

**Review by crcrpar — APPROVED** [review]

**Pull request merged**